File: Android/app/src/main/java/ru/inwords/inwords/core/SimpleDiffUtilCallback.java
package ru.inwords.inwords.core;

import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

public abstract class SimpleDiffUtilCallback<T> extends DiffUtil.Callback {

    private final List<T> oldItems;
    private final List<T> newItems;

    public SimpleDiffUtilCallback(List<T> oldItems, List<T> newItems) {
        this.oldItems = oldItems;
        this.newItems = newItems;
    }

    public abstract boolean areItemsTheSame(T oldItem, T newItem);

    protected abstract boolean areContentsTheSame(T oldItem, T newItem);

    @Override
    public final int getOldListSize() {
        return oldItems.size();
    }

    @Override
    public final int getNewListSize() {
        return newItems.size();
    }

    @Override
    public final boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        T oldItem = oldItems.get(oldItemPosition);
        T newItem = newItems.get(newItemPosition);
        return areItemsTheSame(oldItem, newItem);
    }

    @Override
    public final boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        T oldItem = oldItems.get(oldItemPosition);
        T newItem = newItems.get(newItemPosition);
        return areContentsTheSame(oldItem, newItem);
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/core/SimpleDiffUtilCallback.kt
package ru.inwords.inwords.core

import androidx.recyclerview.widget.DiffUtil

abstract class SimpleDiffUtilCallback<T : Any>(private val oldItems: List<T>, private val newItems: List<T>) : DiffUtil.Callback() {

    abstract fun areItemsTheSame(oldItem: T, newItem: T): Boolean

    protected abstract fun areContentsTheSame(oldItem: T, newItem: T): Boolean

    override fun getOldListSize() = oldItems.size

    override fun getNewListSize() = newItems.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldItems[oldItemPosition]
        val newItem = newItems[newItemPosition]
        return areItemsTheSame(oldItem, newItem)
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldItems[oldItemPosition]
        val newItem = newItems[newItemPosition]
        return areContentsTheSame(oldItem, newItem)
    }
}

class BasicDiffUtilCallback<T : Any> private constructor(old: List<T>, new: List<T>) : SimpleDiffUtilCallback<T>(old, new) {
    override fun areItemsTheSame(oldItem: T, newItem: T): Boolean {
        return oldItem == newItem
    }

    override fun areContentsTheSame(oldItem: T, newItem: T): Boolean {
        return oldItem == newItem
    }

    companion object {
        fun <T : Any> create(old: List<T>, new: List<T>): BasicDiffUtilCallback<T> {
            return BasicDiffUtilCallback(old, new)
        }
    }
}